README.md
# Thailand Election'62 Zone and Candidate Dataset and Browsing Website

![](screenshots/front-pages.png)

## How did we get candidate data?
In the beginning, we had only [PDFs](https://www.ect.go.th/ect_th/news_all.php?cid=165) from The Election Commission of Thailand (ECT). So, we had to parse them manually.
Details about parsing can be founded in `./notebooks/dev.ipyndb`. The process caused several issues, for example, typos from tone marks not ordered properly.

Around 20 Feb 2019, ECT released [SmartVote][smartvote] application, we then tracked down APIs that the application uses and were able to retreive the candidate and party information. 

Below are the datasets extracted from SmartVote's APIs.
1. Zone candidate information 
   ```./data/candidates.csv, ./data/detailed-candidates.csv```
2. Party-list candidate information
   ```./data/partylist_candidates.csv``` with help from [@c4n](//github.com/c4n)
3. Party information, including leader and secretary's names, address, logo, address, policy 
   ```/data/party-information.json, ./data/party-logos```

Details about tracking down API endpoints used by SmartVote will be written in TODO.

## Zone & Candidate Browsering Website
This website has no database. 
This website provides a quick and simple way for users to see who are candidates in their area, using postcode. As zone and candidate data is rarely changed, a static website is sufficient to this purpose without the use of backend and database. Apart from few components needed to be developed, the static-site approach can be massively scaled, for example deployed behind a CDN service. At peak, we had 5,000 concurrent users without any failure. This technique is also adopted by [ELECT Live](https://github.com/codeforthailand/election-live).

This website has two components, front page, zone and party pages. The first page has a form for entering a postcode or address. We use React for HTML DOM manipulation. Zone and party pages are static files that are generated accordingly. Please see `generate-site.py` for more details.

### Development
Every change made to those pagaes, located in `election62`, requires regeneration. One can use the following command to automatically update the site:
```
watch -n 10 'WEBPATH=127.0.0.1:8000 python generate-site.py --env dev'
```

The script generates the site to `./dist`. You can run a static web server from the root directory and access it via `http://<WEBPATH>/dist`. You might encouter cross-origin issues. If that the case, please verify the value of `WEBPATH`.

For example, one can use Python's web server module for this purpose:
```
python -m SimpleHTTPServer
```

### Statistics
![](./screenshots/stats.png)

## **Analysis**
### The presence of political families in 2019 general election
This Jupyter notebook investigates families with multiple MP candidates as well as their affiliation/electoral area in the 2019 Thai general election.  [[Notebook]](notebooks/political-dynasty.ipynb)

# Acknowledgement
This project is published as a part of [ELECT](https://elect.in.th). Founded by several parties who care about democracy and transparent information, the mission of ELECT is to provide insightful information for Thai peoples in order to make a better voting decision. We would like to thank all volunteers behind ELECT, including The Matter, Boonmee Lab, and Minimore, and  all contributors on GitHub.

[mitmproxy]: https://mitmproxy.org
[smartvote]: https://itunes.apple.com/th/app/smart-vote/id1447199802?l=th&mt=8

CHANGELOG.md
# Changelog
- 03/03/19
```
- เพิ่มnotebook สำรวจนามสกุล/ตระกูลที่มีผู้สมัครส.ส.มากกว่าหนึ่งคนขึ้นไป
```
- 25/02/19
```
- ย้ายลิงค์ฟอร์มแจ้ง
- แก้โลโก้ พรรค เพื่อชาติ
```
- 22/02/19
```
- รหัสที่มีปัญหา 33150, 41000, 12120
- https://www.facebook.com/electinth/posts/371876360297770?comment_id=372263686925704&comment_tracking=%7B%22tn%22%3A%22R9%22%7D
- เพิ่ม logo พรรค
```
- 17/02/19
```
# v0.2.1
- search popup
- correct อนาคตใหม่ ดาต้า (กทม.​ 12, 30)
- update หมายเหตุ

# v0.2
- Attach index url to the ELECT logo (Issue#1)
- Use React Prod files (PR#2)
- Track candidate search to Google Analytics.
- Correct อนาคตไทย candidates, not อนาคตใหม่
- logo พรรคกลาง
```

election62/statics/main.js
let CODE_TO_ZONES_PATH = "statics/postcode-to-zones.json";
let TAMBON2ZONE_PATH = "statics/all-tambon-2-election-zones.json";

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = { postcode: undefined, tambon2Zone: undefined, isLoadingTambon2Zone: false };
    this.handlePostcodeChange = this.handlePostcodeChange.bind(this);
    this.handleAddressChange = this.handleAddressChange.bind(this);
  }

  componentDidMount() {
    let self = this;
    let pCodeToZones = fetch(CODE_TO_ZONES_PATH)
      .then(resp => resp.json()) 
      .then(data => {
        self.setState({ codeToZones: data});
      });
  }

  handleAddressChange(event) {
    let v = event.target.value;
    if(!this.state.tambon2Zone && !this.state.isLoadingTambon2Zone){
      this.setState({isLoadingTambon2Zone: true})
      console.log('load tamboon 2 zone')
      let pTambon2Zone = fetch(TAMBON2ZONE_PATH)
        .then(resp => resp.json())
        .then(data => {
          data = data.map(a => {
            a['slug'] = a['tambon'] + '-' + a['amphur'] + '-' + a['province']
            return a
          });
          this.setState({ tambon2Zone: data, isLoadingTambon2Zone: false});
        });
    } else if(this.state.tambon2Zone && !this.state.isLoadingTambon2Zone && v.length > 3) {
      
      let areas = this.state.tambon2Zone.filter(a => a['slug'].indexOf(v) > -1)
        .map(a => {
          a['idf'] = a['slug'].indexOf(v)
          if(a['province'] == 'กรุงเทพมหานคร') {
            a['prefix']  = {
              area: 'เขต',
              subarea: 'แขวง'
            }
          } else {
            a['prefix']  = {
              area: 'อำเภอ',
              subarea: 'ตำบล'
            }
          }
          return a
        })
        .slice(0, 5)

      areas.sort( (a, b) => a['idf'] - b['idf'])
      this.setState({addressQueryAreas: areas, postcode: undefined});
    } else {
      console.log('is loading');
    }

    if(v.length ==0){
      this.setState({addressQueryAreas: undefined});
    }
  }

  handlePostcodeChange(event) {
    let v = event.target.value;

    if (v.length == 5) {
      let z = this.state.codeToZones[v];
      this.setState({ postcode: z, addressQueryAreas: undefined});
    } else {
      this.setState({ postcode: undefined });
    }

  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.postcode && this.state.postcode.zones.length > 0) {
      if(!prevState.postcode ||
        !prevState.postcode.code ||
        (this.state.postcode.code !== prevState.postcode.code))
      gtag('event', 'input', {
        'event_category': 'postcode',
        'event_label': this.state.postcode.code,
        'transport_type': 'beacon'
      });
    }
  }

  render() {
    let zones = [];

    if (this.state.postcode) {
      zones = this.state.postcode.zones.map(z => {
        let key = z.province + "-" + z.zone;
        let link = "z/" + key + ".html";
        return (
          <li key={key} className="zone">
            <a href={link} key={key} target="_blank">
              <div className="header">
                {z.province} เขตเลือกตั้งที่ {z.zone}
              </div>
              <ul>
                {z.areas.map(a => {
                  return (
                    <li className="sub-area" key={a.area}>
                      <span className="area-name">
                        {z.prefixes.area} {a.area}
                      </span>
                      {a.interior.length > 0 && (
                        <span>
                          <b>เฉพาะ </b>
                          {z.prefixes.sub_area} {a.interior.join(", ")}
                        </span>
                      )}
                      {a.exterior.length > 0 && (
                        <span>
                          <b>ยกเว้น </b>
                          {z.prefixes.sub_area} {a.exterior.join(", ")}
                        </span>
                      )}
                      {a.subinterior.length > 0 && (
                        <span>
                          <b>เฉพาะ </b>
                          {a.subinterior.join(", ")}
                        </span>
                      )}
                    </li>
                  );
                })}
              </ul>
            </a>
          </li>
        );
      });
    } else if(this.state.addressQueryAreas){
      zones = this.state.addressQueryAreas.map(z => {
        let key = z.province + "-" + z.zone;
        let link = "z/" + key + ".html";
        return (
          <li key={z.slug} className="zone">
            <a href={link} target="_blank">
              <div className="header">
                {z.province} เขตเลือกตั้งที่ {z.zone}
              </div>
              <span className="area-name">
                {z.prefix.subarea} {z.tambon} {z.prefix.area} {z.amphur}
              </span>
            </a>
          </li>
        );
      });
    }

    return (
      <div>
        <div className="form">
          <div className="postcode-label">
            ค้นหาจาก<b>รหัสไปรษณีย์</b>
          </div>
          <input
            type="text"
            pattern="\d*"
            maxLength="5"
            onChange={this.handlePostcodeChange}
          />
        </div>
        { !this.state.postcode && 
          <div>
            <div className="form-separator">
              <span>หรือ</span>
            </div>
            <div className="form">
              <div className="postcode-label">
                ค้นหาจาก<b>ที่อยู่</b>
              </div>
              <input
                type="text"
                onChange={this.handleAddressChange}
              />
            </div>
          </div>
        }
        {zones.length > 0 && (
          <div className="results">
            <div className="result-label">
            { this.state.postcode && 
              <span>
                มีทั้งหมด {zones.length} เขตเลือกตั้ง
              </span>
            }
            { this.state.addressQueryAreas && 
              <span>
               {zones.length} ที่อยู่ที่ใกล้เคียงคำค้นหา
              </span>
            }
            </div>
            <ul className="zones">
              {zones}
            </ul>
            <div className="ui-guide">
              <i className="fas fa-hand-point-up"></i>
              คลิก เพื่อดูรายชื่อผู้สมัคร 
            </div>
          </div>
        )}
      </div>
    );
  }
}

ReactDOM.render(<Form />, document.getElementById("root"));


generate-site.py
import jinja2
import logging
import shutil
import os
import json
import pandas as pd
from distutils.dir_util import copy_tree
import argparse
import datetime


logging.basicConfig(format='%(asctime)s| %(message)s', level=logging.INFO)

DEST_LOCATION = './dist'
config = {
    'dev': {
        'path': '%s/dist' % os.environ.get('WEBPATH', 'localhost:8000')
    },
    'stage': {
        'path': 'pat.chormai.org/election-62-staging'  
    },
    'prod': {
        'path': 'elect.in.th/candidates'  
    }
}


parser = argparse.ArgumentParser(description='Generate election 2556 candidate sites')
parser.add_argument('--env', dest='environment', help='environment to generate')

args = parser.parse_args()


deploy_path = config[args.environment]['path']

def prepend_dir(path):
    return '%s/%s' % (DEST_LOCATION, path)

def cache_buster_url(url, cache_key=datetime.datetime.today().strftime('%Y-%m-%d-%H-%M')):
    return '%s?v=%s' % (url, cache_key)

if os.path.exists(DEST_LOCATION):
    shutil.rmtree(DEST_LOCATION)
os.mkdir(DEST_LOCATION)

copy_tree('./election62/statics', prepend_dir('statics'))

templateEnv = jinja2.Environment(
    extensions=['jinja2_time.TimeExtension'],
    loader=jinja2.PackageLoader('election62', 'templates')
)

templateEnv.filters['cache_buster'] = cache_buster_url

logging.info('Generating files for %s' % config[args.environment]['path'])

# render frontpage
logging.info('Rendering index.html')
template = templateEnv.get_template('index.html')
template.stream(deploy_path=deploy_path, env=args.environment).dump(prepend_dir('index.html'))

# render zone pages
logging.info('Rendering zone pages')
with open('./data/election-zones.json') as ez:  
    election_zones = json.load(ez)

with open('./data/party-abbrevations.json') as pe:  
    party2abbv = dict(map(lambda x: (x['name'], x['id']), json.load(pe)))
    
df_candidates = pd.read_csv('./data/detailed-candidates.csv')
df_candidates['party_abbv'] = df_candidates.PartyName.apply(lambda x: party2abbv.get(x, None))

os.mkdir(prepend_dir('z'))

template = templateEnv.get_template('zone.html')
for ez in election_zones:
    candidates = df_candidates[
            (df_candidates['province_name'] == ez['province']) &
            (df_candidates['zone_number'] == ez['zone'])
        ].to_dict('records')

    filename = '%s-%s' % (ez['province'], ez['zone'])
    if len(candidates) < 10:
        logging.info('Zone %s has <10 candidates, skip it!' % filename)
    else:
        candidates = sorted(candidates, key=lambda x: x['CandidateNo'])
        template.stream(deploy_path=deploy_path, zone=ez, candidates=candidates, env=args.environment) \
            .dump(prepend_dir('z/%s.html' % filename))

# render party pages
logging.info('Rendering party pages')
os.mkdir(prepend_dir('p'))

template = templateEnv.get_template('party.html')

with open('./data/party-information.json') as pt:  
    party_info = json.load(pt)

for p in party_info:
    filename = p['slug']
    p['party_abbv'] = party2abbv.get(p['slug'], None)

    template.stream(deploy_path=deploy_path, env=args.environment, party=p) \
        .dump(prepend_dir('p/%s.html' % filename))